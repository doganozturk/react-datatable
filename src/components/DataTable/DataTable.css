.DataTable h1 {
  font-size: 30px;
  color: #fff;
  text-transform: uppercase;
  font-weight: 300;
  text-align: center;
  margin-bottom: 15px;
}

.DataTable table {
  width: 100%;
  table-layout: fixed;
}

.DataTable .tbl-header {
  background-color: rgba(255, 255, 255, 0.3);
}

.DataTable .tbl-content {
  margin-top: 0;
  border: 1px solid rgba(255, 255, 255, 0.3);
}

.DataTable th {
  padding: 20px 15px;
  text-align: left;
  font-weight: 500;
  font-size: 12px;
  color: #fff;
  text-transform: uppercase;
}
