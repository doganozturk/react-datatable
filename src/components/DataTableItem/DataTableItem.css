.DataTableItem td {
  padding: 15px;
  text-align: left;
  vertical-align: middle;
  font-weight: 300;
  font-size: 12px;
  color: #fff;
  border-bottom: solid 1px rgba(255, 255, 255, 0.1);
}
