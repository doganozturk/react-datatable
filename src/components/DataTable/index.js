import React, { Component } from 'react';
import items from '../../data/data.json';
import './DataTable.css';
import DataTableItem from '../DataTableItem';

class DataTable extends Component {
  constructor(props) {
    super(props);

    this.state = {
      items: this.listToTree(items)
    };
  }

  listToTree(data, { idKey, parentKey, childrenKey } = {}) {
    const ID_KEY = idKey || 'ID';
    const PARENT_KEY = parentKey || 'parentID';
    const CHILDREN_KEY = childrenKey || 'children';

    const tree = [];
    const childrenOf = {};

    let item;
    let id;
    let parentId;

    for (let i = 0, length = data.length; i < length; i++) {
      item = data[i];
      id = item[ID_KEY];
      parentId = item[PARENT_KEY] || 0;
      // every item may have children
      childrenOf[id] = childrenOf[id] || [];
      // init its children
      item[CHILDREN_KEY] = childrenOf[id];

      if (parentId !== 0) {
        // init its parent's children object
        childrenOf[parentId] = childrenOf[parentId] || [];
        // push it into its parent's children object
        childrenOf[parentId].push(item);
      } else {
        tree.push(item);
      }
    }
    console.log(tree);
    return tree;
  }

  renderItems() {
    return this.state.items.map(item =>
      <DataTableItem item={item} key={item.ID} />
    );
  }

  render() {
    return (
      <div className="DataTable">
        <h1>Fixed Table header</h1>
        <div className="tbl-header">
          <table cellPadding="0" cellSpacing="0">
            <thead>
              <tr>
                <th>ID</th>
                <th>Phone</th>
                <th>City</th>
                <th>Name</th>
              </tr>
            </thead>
          </table>
        </div>
        <div className="tbl-content">
          <table cellPadding="0" cellSpacing="0">
            <tbody>
              {this.renderItems()}
            </tbody>
          </table>
        </div>
      </div>
    );
  }
}

export default DataTable;
