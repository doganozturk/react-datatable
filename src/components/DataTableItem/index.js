import React, { Component } from 'react';
import './DataTableItem.css';

class DataTableItem extends Component {
  render() {
    return (
      <tr className="DataTableItem">
        <td>
          {this.props.item.ID}
        </td>
        <td>
          {this.props.item.parentID}
        </td>
        <td>
          {this.props.item.Phone}
        </td>
        <td>
          {this.props.item.City}
        </td>
        <td>
          {this.props.item.Name}
        </td>
      </tr>
    );
  }
}

export default DataTableItem;
