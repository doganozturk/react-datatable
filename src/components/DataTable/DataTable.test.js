import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import DataTable from './';
import DataTableItem from '../DataTableItem';

describe(DataTable, () => {
  const component = shallow(<DataTable />);

  it('renders and matches our snapshot', () => {
    const component = renderer.create(<DataTable />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('contains the same number of DataTableItem components as items', () => {
    const dataTableItems = component.find(DataTableItem).length;
    const items = component.state('items').length;
    expect(dataTableItems).toEqual(items);
  });

  it('removes an item from items when deleteItem function is called', () => {
    const before = component.find(DataTableItem).length;
    component.instance().deleteItem(1);
    const after = component.find(DataTableItem).length;
    expect(before).toBeGreaterThan(after);
  });
});
