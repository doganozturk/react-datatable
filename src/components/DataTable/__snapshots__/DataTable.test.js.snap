// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class DataTable extends _react.Component {
  constructor(props) {
    super(props);

    this.state = {
      items: this.listToTree(_data2.default) };

  }

  listToTree(data, { idKey, parentKey, childrenKey } = {}) {
    const ID_KEY = idKey || 'ID';
    const PARENT_KEY = parentKey || 'parentID';
    const CHILDREN_KEY = childrenKey || 'children';

    const tree = [];
    const childrenOf = {};

    let item;
    let id;
    let parentId;

    for (let i = 0, length = data.length; i < length; i++) {
      item = data[i];
      id = item[ID_KEY];
      parentId = item[PARENT_KEY] || 0;
      // every item may have children
      childrenOf[id] = childrenOf[id] || [];
      // init its children
      item[CHILDREN_KEY] = childrenOf[id];

      if (parentId !== 0) {
        // init its parent's children object
        childrenOf[parentId] = childrenOf[parentId] || [];
        // push it into its parent's children object
        childrenOf[parentId].push(item);
      } else {
        tree.push(item);
      }
    }
    console.log(tree);
    return tree;
  }

  renderItems() {
    return this.state.items.map(item =>
    _react2.default.createElement(_DataTableItem2.default, { item: item, key: item.ID, __source: { fileName: _jsxFileName, lineNumber: 51 }, __self: this }));

  }

  render() {
    return (
      _react2.default.createElement('div', { className: 'DataTable', __source: { fileName: _jsxFileName, lineNumber: 57 }, __self: this },
        _react2.default.createElement('h1', { __source: { fileName: _jsxFileName, lineNumber: 58 }, __self: this }, 'Fixed Table header'),
        _react2.default.createElement('div', { className: 'tbl-header', __source: { fileName: _jsxFileName, lineNumber: 59 }, __self: this },
          _react2.default.createElement('table', { cellPadding: '0', cellSpacing: '0', __source: { fileName: _jsxFileName, lineNumber: 60 }, __self: this },
            _react2.default.createElement('thead', { __source: { fileName: _jsxFileName, lineNumber: 61 }, __self: this },
              _react2.default.createElement('tr', { __source: { fileName: _jsxFileName, lineNumber: 62 }, __self: this },
                _react2.default.createElement('th', { __source: { fileName: _jsxFileName, lineNumber: 63 }, __self: this }, 'ID'),
                _react2.default.createElement('th', { __source: { fileName: _jsxFileName, lineNumber: 64 }, __self: this }, 'Phone'),
                _react2.default.createElement('th', { __source: { fileName: _jsxFileName, lineNumber: 65 }, __self: this }, 'City'),
                _react2.default.createElement('th', { __source: { fileName: _jsxFileName, lineNumber: 66 }, __self: this }, 'Name'))))),




        _react2.default.createElement('div', { className: 'tbl-content', __source: { fileName: _jsxFileName, lineNumber: 71 }, __self: this },
          _react2.default.createElement('table', { cellPadding: '0', cellSpacing: '0', __source: { fileName: _jsxFileName, lineNumber: 72 }, __self: this },
            _react2.default.createElement('tbody', { __source: { fileName: _jsxFileName, lineNumber: 73 }, __self: this },
              this.renderItems())))));





  }} renders and matches our snapshot 1`] = `
<div
  className="DataTable"
>
  <h1>
    Fixed Table header
  </h1>
  <div
    className="tbl-header"
  >
    <table
      cellPadding="0"
      cellSpacing="0"
    >
      <thead>
        <tr>
          <th>
            ID
          </th>
          <th>
            Phone
          </th>
          <th>
            City
          </th>
          <th>
            Name
          </th>
        </tr>
      </thead>
    </table>
  </div>
  <div
    className="tbl-content"
  >
    <table
      cellPadding="0"
      cellSpacing="0"
    >
      <tbody>
        <tr
          className="DataTableItem"
        >
          <td>
            1
          </td>
          <td>
            (403) 125-2552
          </td>
          <td>
            Coevorden
          </td>
          <td>
            Grady
          </td>
        </tr>
        <tr
          className="DataTableItem"
        >
          <td>
            2
          </td>
          <td>
            (979) 486-1932
          </td>
          <td>
            Chełm
          </td>
          <td>
            Scarlet
          </td>
        </tr>
        <tr
          className="DataTableItem"
        >
          <td>
            3
          </td>
          <td>
            (573) 685-8350
          </td>
          <td>
            Wardha
          </td>
          <td>
            Adria
          </td>
        </tr>
        <tr
          className="DataTableItem"
        >
          <td>
            5
          </td>
          <td>
            (755) 968-6539
          </td>
          <td>
            Gönen
          </td>
          <td>
            Madeson
          </td>
        </tr>
        <tr
          className="DataTableItem"
        >
          <td>
            7
          </td>
          <td>
            (896) 297-6568
          </td>
          <td>
            Louvain-la-Neuve
          </td>
          <td>
            Celeste
          </td>
        </tr>
        <tr
          className="DataTableItem"
        >
          <td>
            10
          </td>
          <td>
            (450) 579-0491
          </td>
          <td>
            Middelburg
          </td>
          <td>
            Madaline
          </td>
        </tr>
        <tr
          className="DataTableItem"
        >
          <td>
            11
          </td>
          <td>
            (111) 162-2502
          </td>
          <td>
            Birecik
          </td>
          <td>
            Chandler
          </td>
        </tr>
        <tr
          className="DataTableItem"
        >
          <td>
            17
          </td>
          <td>
            (407) 519-9894
          </td>
          <td>
            Roubaix
          </td>
          <td>
            Justine
          </td>
        </tr>
        <tr
          className="DataTableItem"
        >
          <td>
            21
          </td>
          <td>
            (601) 944-5214
          </td>
          <td>
            Pointe-au-Pic
          </td>
          <td>
            Iris
          </td>
        </tr>
        <tr
          className="DataTableItem"
        >
          <td>
            25
          </td>
          <td>
            (964) 413-7033
          </td>
          <td>
            Joliet
          </td>
          <td>
            Leo
          </td>
        </tr>
        <tr
          className="DataTableItem"
        >
          <td>
            60
          </td>
          <td>
            (397) 265-7757
          </td>
          <td>
            Ahmadpur East
          </td>
          <td>
            Zelenia
          </td>
        </tr>
        <tr
          className="DataTableItem"
        >
          <td>
            68
          </td>
          <td>
            (807) 949-9083
          </td>
          <td>
            Villers-la-Bonne-Eau
          </td>
          <td>
            Wyatt
          </td>
        </tr>
        <tr
          className="DataTableItem"
        >
          <td>
            69
          </td>
          <td>
            (515) 371-8783
          </td>
          <td>
            Poggiodomo
          </td>
          <td>
            Aidan
          </td>
        </tr>
        <tr
          className="DataTableItem"
        >
          <td>
            72
          </td>
          <td>
            (134) 258-4612
          </td>
          <td>
            Kukatpalle
          </td>
          <td>
            Morgan
          </td>
        </tr>
        <tr
          className="DataTableItem"
        >
          <td>
            77
          </td>
          <td>
            (430) 230-5837
          </td>
          <td>
            Bregenz
          </td>
          <td>
            Ethan
          </td>
        </tr>
        <tr
          className="DataTableItem"
        >
          <td>
            78
          </td>
          <td>
            (953) 595-4023
          </td>
          <td>
            Shahjahanpur
          </td>
          <td>
            Madeline
          </td>
        </tr>
        <tr
          className="DataTableItem"
        >
          <td>
            80
          </td>
          <td>
            (546) 698-1207
          </td>
          <td>
            Arnhem
          </td>
          <td>
            Quinlan
          </td>
        </tr>
        <tr
          className="DataTableItem"
        >
          <td>
            84
          </td>
          <td>
            (861) 753-8646
          </td>
          <td>
            Wolfsburg
          </td>
          <td>
            Upton
          </td>
        </tr>
        <tr
          className="DataTableItem"
        >
          <td>
            97
          </td>
          <td>
            (636) 383-2370
          </td>
          <td>
            St. Catharines
          </td>
          <td>
            Maxwell
          </td>
        </tr>
        <tr
          className="DataTableItem"
        >
          <td>
            99
          </td>
          <td>
            (336) 239-7842
          </td>
          <td>
            Goderich
          </td>
          <td>
            Byron
          </td>
        </tr>
        <tr
          className="DataTableItem"
        >
          <td>
            100
          </td>
          <td>
            (714) 477-8291
          </td>
          <td>
            St. Ives
          </td>
          <td>
            Bryar
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
